add user_pin_length and user_pin_description to Credential Offer (#95)
openid/OpenID4VCI: openid-4-verifiable-credential-issuance-1_0.md

  * `pre-authorized_code`: REQUIRED. The code representing the Credential Issuer's authorization for the Wallet to obtain Credentials of a certain type. This code MUST be short lived and single use. If the Wallet decides to use the Pre-Authorized Code Flow, this parameter value MUST be included in the subsequent Token Request with the Pre-Authorized Code Flow.
  * `user_pin_required`: OPTIONAL. Boolean value specifying whether the AS expects presentation of a user PIN along with the Token Request in a Pre-Authorized Code Flow. Default is `false`. This PIN is intended to bind the Pre-Authorized Code to a certain transaction to prevent replay of this code by an attacker that, for example, scanned the QR code while standing behind the legitimate user. It is RECOMMENDED to send the PIN via a separate channel. If the Wallet decides to use the Pre-Authorized Code Flow, the PIN value MUST be sent in the `user_pin` parameter with the respective Token Request.
  * `user_pin_length`: OPTIONAL. JSON Integer specifying the length of the PIN. This helps the Wallet to render the PIN screen and improve the user experience. When `user_pin_required` is `false`, this parameter MUST NOT be present.

>   * `user_pin_description`: OPTIONAL. JSON String containing guidance for the Holder of the Wallet how to obtain the PIN, e.g. specifying over which communication channel the PIN is delivered. The Wallet is RECOMMENDED to display this description next to the PIN entry form to improve the user experience. The length of the JSON String MUST be limited to 300 characters. The `user_pin_description` does not support internationalization (i18n), however the Issuer may detect the Holder's language by previous communication or an HTTP Accept-Language header within an HTTP GET request for a  `credential_offer_uri`. When `user_pin_required` is `false`, this parameter MUST NOT be present.

Reviewer: Detail: "JSON String containing guidance for the Holder of the Wallet *on* how to obtain the PIN".

Author: applied

  * `interval`: OPTIONAL. The minimum amount of time in seconds that the Wallet SHOULD wait between polling requests to the token endpoint (in case the Authorization Server responds with error code `authorization_pending` - see (#token_error_response)). If no value is provided, Wallets MUST use 5 as the default.

> The following non-normative example shows a Credential Offer object where the Credential Issuer offers the issuance of two Credentials of different formats, one as JSON string ("UniversityDegree_JWT") and the other one as JSON object:

Reviewer 2: ```suggestion
The following non-normative example shows a Credential Offer object where the Credential Issuer offers the issuance of two Credentials of different formats, one as a string ("UniversityDegree_JWT") and the other one as an object:
```

* Grant Type `urn:ietf:params:oauth:grant-type:pre-authorized_code`:
  * `pre-authorized_code`: REQUIRED. The code representing the Credential Issuer's authorization for the Wallet to obtain Credentials of a certain type. This code MUST be short lived and single use. If the Wallet decides to use the Pre-Authorized Code Flow, this parameter value MUST be included in the subsequent Token Request with the Pre-Authorized Code Flow.
  * `user_pin_required`: OPTIONAL. Boolean value specifying whether the AS expects presentation of a user PIN along with the Token Request in a Pre-Authorized Code Flow. Default is `false`. This PIN is intended to bind the Pre-Authorized Code to a certain transaction to prevent replay of this code by an attacker that, for example, scanned the QR code while standing behind the legitimate user. It is RECOMMENDED to send the PIN via a separate channel. If the Wallet decides to use the Pre-Authorized Code Flow, the PIN value MUST be sent in the `user_pin` parameter with the respective Token Request.

>   * `user_pin_length`: OPTIONAL. JSON Integer specifying the length of the PIN. This helps the Wallet to render the PIN screen and improve the user experience. When `user_pin_required` is `false`, this parameter MUST NOT be present.
>   * `user_pin_description`: OPTIONAL. JSON String containing guidance for the Holder of the Wallet how to obtain the PIN, e.g. specifying over which communication channel the PIN is delivered. The Wallet is RECOMMENDED to display this description next to the PIN entry form to improve the user experience. The length of the JSON String MUST be limited to 300 characters. The `user_pin_description` does not support internationalization (i18n), however the Issuer may detect the Holder's language by previous communication or an HTTP Accept-Language header within an HTTP GET request for a  `credential_offer_uri`. When `user_pin_required` is `false`, this parameter MUST NOT be present.

Reviewer 2: ```suggestion
  * `user_pin_length`: OPTIONAL. Integer specifying the length of the PIN. This helps the Wallet to render the PIN screen and improve the user experience. When `user_pin_required` is `false`, this parameter MUST NOT be present.
  * `user_pin_description`: OPTIONAL. String containing guidance for the Holder of the Wallet how to obtain the PIN, e.g., specifying over which communication channel the PIN is delivered. The Wallet is RECOMMENDED to display this description next to the PIN entry form to improve the user experience. The length of the JSON String MUST be limited to 300 characters. The `user_pin_description` does not support internationalization (i18n), however the Issuer may detect the Holder's language by previous communication or an HTTP Accept-Language header within an HTTP GET request for a  `credential_offer_uri`. When `user_pin_required` is `false`, this parameter MUST NOT be present.
```

Author: While I don't object to this, the specification is not very consistent on whether to name things "String" or "JSON String". Is there a consensus on this?

Reviewer 2: My point of view is that it should be avoided to say "JSON String" (because the encoding is defined on the outer object and here we're just talking about the data structure).

Reviewer 3: Note that the 'JSON String' and 'JSON Integer' language are used throughout OpenID Connect and https://datatracker.ietf.org/doc/html/rfc7519 etc.

We should probably agree on the policy and use that policy consistently rather than changing specific instances in unrelated pull requests.

Reviewer 2: I'll open an issue!

Reviewer 2: Since we agreed in #97 to drop "JSON", I have reopened this suggestion.

Author: applied

>   * `user_pin_description`: OPTIONAL. JSON String containing guidance for the Holder of the Wallet how to obtain the PIN, e.g. specifying over which communication channel the PIN is delivered. The Wallet is RECOMMENDED to display this description next to the PIN entry form to improve the user experience. The length of the JSON String MUST be limited to 300 characters. The `user_pin_description` does not support internationalization (i18n), however the Issuer may detect the Holder's language by previous communication or an HTTP Accept-Language header within an HTTP GET request for a  `credential_offer_uri`. When `user_pin_required` is `false`, this parameter MUST NOT be present.

Author: applied

  * `authorization_server`: OPTIONAL string that the Wallet can use to identify the Authorization Server to use with this grant type when `authorization_servers` parameter in the Credential Issuer metadata has multiple entries. MUST NOT be used otherwise. The value of this parameter MUST match with one of the values in the `authorization_servers` array obtained from the Credential Issuer metadata.
* Grant Type `urn:ietf:params:oauth:grant-type:pre-authorized_code`:
  * `pre-authorized_code`: REQUIRED. The code representing the Credential Issuer's authorization for the Wallet to obtain Credentials of a certain type. This code MUST be short lived and single use. If the Wallet decides to use the Pre-Authorized Code Flow, this parameter value MUST be included in the subsequent Token Request with the Pre-Authorized Code Flow.

>   * `tx_code`: OPTIONAL. An object specifying whether the AS expects presentation of a Transaction Code by the End-User along with the Token Request in a Pre-Authorized Code Flow. If the AS does not expect a Transaction Code, this object is absent, this is the default. The Transaction Code is intended to bind the Pre-Authorized Code to a certain transaction to prevent replay of this code by an attacker that, for example, scanned the QR code while standing behind the legitimate End-User. It is RECOMMENDED to send the Transaction Code via a separate channel. If the Wallet decides to use the Pre-Authorized Code Flow, the Transaction Code value MUST be sent in the `tx_code` parameter with the respective Token Request. If no `length` or `description` is given, this object may be empty, indicating that a Transaction Code is required.

Reviewer 4: ```suggestion
  * `tx_code`: OPTIONAL. An object specifying whether the AS expects presentation of a Transaction Code by the End-User along with the Token Request in a Pre-Authorized Code Flow. If the AS does not expect a Transaction Code, this object is absent, this is the default. The Transaction Code is intended to bind the Pre-Authorized Code to a certain transaction to prevent replay of this code by an attacker that, for example, scanned the QR code while standing behind the legitimate End-User. It is RECOMMENDED to send the Transaction Code via a separate channel. If the Wallet decides to use the Pre-Authorized Code Flow, the Transaction Code value MUST be sent in the `tx_code` parameter with the respective Token Request as defined in (#token_request). If no `length` or `description` is given, this object may be empty, indicating that a Transaction Code is required.
```

    * `length`: OPTIONAL. Integer specifying the length of the Transaction Code. This helps the Wallet to render the input screen and improve the user experience.

>     * `description`: OPTIONAL. String containing guidance for the Holder of the Wallet on how to obtain the Transaction Code, e.g. describing over which communication channel it is delivered. The Wallet is RECOMMENDED to display this description next to the Transaction Code input screen to improve the user experience. The length of the String MUST NOT exceed 300 characters. The `description` does not support internationalization (i18n), however the Issuer MAY detect the Holder's language by previous communication or an HTTP Accept-Language header within an HTTP GET request for a  `credential_offer_uri`.

Reviewer 4: ```suggestion
    * `description`: OPTIONAL. String containing guidance for the Holder of the Wallet on how to obtain the Transaction Code, e.g. describing over which communication channel it is delivered. The Wallet is RECOMMENDED to display this description next to the Transaction Code input screen to improve the user experience. The length of the string MUST NOT exceed 300 characters. The `description` does not support internationalization (i18n), however the Issuer MAY detect the Holder's language by previous communication or an HTTP Accept-Language header within an HTTP GET request for a Credential Offer URI.
```

  * `tx_code`: OPTIONAL. An object specifying whether the AS expects presentation of a Transaction Code by the End-User along with the Token Request in a Pre-Authorized Code Flow. If the AS does not expect a Transaction Code, this object is absent, this is the default. The Transaction Code is intended to bind the Pre-Authorized Code to a certain transaction to prevent replay of this code by an attacker that, for example, scanned the QR code while standing behind the legitimate End-User. It is RECOMMENDED to send the Transaction Code via a separate channel. If the Wallet decides to use the Pre-Authorized Code Flow, the Transaction Code value MUST be sent in the `tx_code` parameter with the respective Token Request as defined in (#token_request). If no `length` or `description` is given, this object may be empty, indicating that a Transaction Code is required.

>     * `input_mode` : OPTIONAL. String specifying the input characters set. Possible values are `numeric` (only digits) and `text` (any character). The default is `numeric`.

Reviewer 4: should this say that this input_mode can also control the keyboard being displayed to the user?

    * `length`: OPTIONAL. Integer specifying the length of the Transaction Code. This helps the Wallet to render the input screen and improve the user experience.

>     * `description`: OPTIONAL. String containing guidance for the Holder of the Wallet on how to obtain the Transaction Code, e.g. describing over which communication channel it is delivered. The Wallet is RECOMMENDED to display this description next to the Transaction Code input screen to improve the user experience. The length of the string MUST NOT exceed 300 characters. The `description` does not support internationalization (i18n), however the Issuer MAY detect the Holder's language by previous communication or an HTTP Accept-Language header within an HTTP GET request for a Credential Offedefiningr URI.

Reviewer 4: ```suggestion
    * `description`: OPTIONAL. String containing guidance for the Holder of the Wallet on how to obtain the Transaction Code, e.g. describing over which communication channel it is delivered. The Wallet is RECOMMENDED to display this description next to the Transaction Code input screen to improve the user experience. The length of the string MUST NOT exceed 300 characters. The `description` does not support internationalization (i18n), however the Issuer MAY detect the Holder's language by previous communication or an HTTP Accept-Language header within an HTTP GET request for a Credential Offer URI.
```

The following are the extension parameters to the Token Request used in a Pre-Authorized Code Flow defined by this specification:

* `pre-authorized_code`: The code representing the authorization to obtain Credentials of a certain type. This parameter MUST be present if the `grant_type` is `urn:ietf:params:oauth:grant-type:pre-authorized_code`.

> * `user_pin`: OPTIONAL. String value containing a user PIN. This value MUST be present if `user_pin` object was present in the Credential Offer. The string value MUST consist of a maximum of eight numeric characters (the numbers 0 - 9). This parameter MUST only be used if the `grant_type` is `urn:ietf:params:oauth:grant-type:pre-authorized_code`.

Reviewer 2: ```suggestion
* `user_pin`: OPTIONAL. String value containing a user PIN. This value MUST be present if a `user_pin` object was present in the Credential Offer (including if the object was empty). The string value MUST consist of a maximum of eight numeric characters (the numbers 0 - 9). This parameter MUST only be used if the `grant_type` is `urn:ietf:params:oauth:grant-type:pre-authorized_code`.
```

   
   -19
  

>    * added `user_pin_description` and `user_pin_length` to Credential Offer

Reviewer 3: ```suggestion
   * replaced `user_pin_required` in Credential Offer with a `user_pin` object that also now contains `description` and `length`
```
